File: backend/src/main/java/com/velocity/map/Pathfinding.java
package com.velocity.map;

import com.velocity.map.MapModels.MapNode;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.PriorityQueue;

/**
 * Dijkstra shortest path over the road graph using a binary-heap priority
 * queue. Costs come from {@link CityMap#effectiveCost}, so live incidents that
 * inflate edge weights automatically push routes around congestion.
 */
@Component
@RequiredArgsConstructor
public class Pathfinding {

    private final CityMap cityMap;

    public record PathResult(List<String> nodeIds, double totalCost, double totalDistance) {
    }

    public PathResult findPath(String startNodeId, String endNodeId) {
        if (startNodeId == null || endNodeId == null) return null;
        if (startNodeId.equals(endNodeId)) {
            return new PathResult(List.of(startNodeId), 0, 0);
        }
        if (cityMap.node(startNodeId) == null || cityMap.node(endNodeId) == null) {
            return null;
        }

        Map<String, Double> dist = new HashMap<>();
        Map<String, String> prev = new HashMap<>();
        PriorityQueue<double[]> heap = new PriorityQueue<>((a, b) -> Double.compare(a[1], b[1]));
        Map<Integer, String> idForEntry = new HashMap<>();

        // double[] {entryId, priority}; map entryId -> nodeId to keep heap numeric
        int[] counter = {0};
        dist.put(startNodeId, 0.0);
        pushEntry(heap, idForEntry, counter, startNodeId, 0.0);

        while (!heap.isEmpty()) {
            double[] top = heap.poll();
            String current = idForEntry.get((int) top[0]);
            double priority = top[1];
            if (current == null) continue;
            if (current.equals(endNodeId)) break;
            if (priority > dist.getOrDefault(current, Double.MAX_VALUE)) continue;

            for (CityMap.Adjacency adj : cityMap.neighbors(current)) {
                double newDist = dist.getOrDefault(current, Double.MAX_VALUE) + cityMap.effectiveCost(adj);
                if (newDist < dist.getOrDefault(adj.nodeId(), Double.MAX_VALUE)) {
                    dist.put(adj.nodeId(), newDist);
                    prev.put(adj.nodeId(), current);
                    pushEntry(heap, idForEntry, counter, adj.nodeId(), newDist);
                }
            }
        }

        if (!prev.containsKey(endNodeId)) {
            return null;
        }

        List<String> path = new ArrayList<>();
        String current = endNodeId;
        while (current != null) {
            path.add(current);
            current = prev.get(current);
        }
        Collections.reverse(path);

        double totalCost = dist.getOrDefault(endNodeId, 0.0);
        double totalDistance = 0;
        for (int i = 0; i < path.size() - 1; i++) {
            totalDistance += cityMap.distanceBetween(path.get(i), path.get(i + 1));
        }
        return new PathResult(path, totalCost, totalDistance);
    }

    private void pushEntry(PriorityQueue<double[]> heap, Map<Integer, String> idForEntry,
                           int[] counter, String nodeId, double priority) {
        int id = counter[0]++;
        idForEntry.put(id, nodeId);
        heap.add(new double[]{id, priority});
    }

    public MapNode findNearestNode(double x, double y) {
        MapNode nearest = null;
        double min = Double.MAX_VALUE;
        for (MapNode n : cityMap.getNodes()) {
            double d = Math.hypot(n.x() - x, n.y() - y);
            if (d < min) {
                min = d;
                nearest = n;
            }
        }
        return nearest;
    }
}

File: engine/src/pathfinding.ts
import type { CityMap } from "./cityMap";

export interface PathResult { nodeIds: string[]; totalCost: number; totalDistance: number; }

/** Binary-heap min-priority queue keyed by numeric priority. */
class MinHeap {
  private a: { id: string; p: number }[] = [];
  get size() { return this.a.length; }
  push(id: string, p: number) {
    this.a.push({ id, p });
    let i = this.a.length - 1;
    while (i > 0) {
      const parent = (i - 1) >> 1;
      if (this.a[parent].p <= this.a[i].p) break;
      [this.a[parent], this.a[i]] = [this.a[i], this.a[parent]];
      i = parent;
    }
  }
  pop(): { id: string; p: number } | undefined {
    const top = this.a[0];
    const last = this.a.pop();
    if (this.a.length && last) {
      this.a[0] = last;
      let i = 0;
      const n = this.a.length;
      for (;;) {
        const l = 2 * i + 1, r = 2 * i + 2;
        let s = i;
        if (l < n && this.a[l].p < this.a[s].p) s = l;
        if (r < n && this.a[r].p < this.a[s].p) s = r;
        if (s === i) break;
        [this.a[s], this.a[i]] = [this.a[i], this.a[s]];
        i = s;
      }
    }
    return top;
  }
}

/**
 * Dijkstra shortest path over the road graph. Costs come from
 * CityMap.effectiveCost, so live incidents that inflate edge weights
 * automatically push routes around congestion.
 */
export function findPath(map: CityMap, start: string, end: string): PathResult | null {
  if (!start || !end) return null;
  if (start === end) return { nodeIds: [start], totalCost: 0, totalDistance: 0 };
  if (!map.node(start) || !map.node(end)) return null;

  const dist = new Map<string, number>([[start, 0]]);
  const prev = new Map<string, string>();
  const heap = new MinHeap();
  heap.push(start, 0);

  while (heap.size) {
    const top = heap.pop()!;
    const cur = top.id;
    if (cur === end) break;
    if (top.p > (dist.get(cur) ?? Infinity)) continue;
    for (const adj of map.neighbors(cur)) {
      const nd = (dist.get(cur) ?? Infinity) + map.effectiveCost(adj);
      if (nd < (dist.get(adj.nodeId) ?? Infinity)) {
        dist.set(adj.nodeId, nd);
        prev.set(adj.nodeId, cur);
        heap.push(adj.nodeId, nd);
      }
    }
  }

  if (!prev.has(end)) return null;
  const path: string[] = [];
  let cur: string | undefined = end;
  while (cur) { path.push(cur); cur = prev.get(cur); }
  path.reverse();

  let totalDistance = 0;
  for (let i = 0; i < path.length - 1; i++) {
    totalDistance += map.distanceBetween(path[i], path[i + 1]);
  }
  return { nodeIds: path, totalCost: dist.get(end) ?? 0, totalDistance };
}
